tsinfer/formats.py

```diff
@@ -1699,6 +1733,7 @@ def add_site(
         metadata=None,
         inference=None,
         time=None,
+        ancestral_state=None,
```

Comment:
I'm not sure tskit.MISSING_DATA is helpful here as an API, as we're mixing strings and integers on input. What if we had a special argument value "unknown" here to mark it? It would make for more readable code I think,
```python
sd.add_site(alleles="ACGT", genotypes=[0, 0, 0, 1], ancestral_state="unknown")
```

since we're not actually storing ``ancestral_state`` there's no advantage to using -1, and (I think) it makes sense to keep the types of the parameter clear.

Reply:
Hmm, isn't it a bit ugly (e.g. what if the allele actually is the string "unknown"!). I would prefer something that can't be interpreted as a string. None would be an obvious one, but we are using that as the default to mean `0`.

Alternatively we could ask for the ancestral state index, and disallow negative values, other than `-1`, but then (as you pointed out above, I think), the user might reasonably expect `-1` to mean the last allele.

Finally, we could ask the user to add `None` to the list of alleles, and indicate that for the index, but that seems a bit convoluted to me.

Reply:
Hah, good point. Hmm, what if we did something like
```python
ANCESTRAL_STATE_IS_FIRST = object() # Standard singleton pattern
...
def add_site(self, ..., ancestral_state=ANCESTRAL_STATE_IS_FIRST):
    if ancestral_state is ANCESTRAL_STATE_IS_FIRST:
        ancestral_state_index = 0
    elif ancestral_state is None:
        ancestral_state_index = -1 # Unknown
    else:
         ancestral_state_index = alleles.index(ancestral_state)
```

The idea is that we're only really supporting old code using the default case, so we can use ``None`` in a more useful way like this.

Thoughts?

Reply:
If we don't mind having something other than `None` as the default, then it might be simpler to have the parameter being the `ancestral_state_index` (default `0`), and have `None` to mean unknown? Then that matches exactly what we store in the SampleData file anyway.

Reply:
Also, if we want a shorter name, rather than use the term `ancestral_state_index`, we could use `ancestral_allele` to mean the allele index, and reserve the term `ancestral_state` (which matches the tskit terminology) to mean the `str` value?

Reply:
I like ``ancestral_allele`` - we generally store indexes and don't put the name on the variable, and this avoids confusion with the tskit 'state'. So let's do that.

I think specifying "ancestral_state" as a string is probably the easiest for users, but then maybe we're over thinking this as the plan is to remove this step of the workflow (add_site) ASAP.

So: let's make the API this:
```python
def add_site(self, ..., ancestral_allele=None):
     if ancestral_allele is None:
          ancestral_allele = 0
     # do the stuff.
```

We can always add another mutually exclusive parameter ``ancestral_state`` later if we want it.

Reply:
OK, great. A minor point though: I thought your idea was to have the default as (explicitly) "0", and to have `None` mean "no ancestral allele known". So:

```python
def add_site(self, ..., ancestral_allele=0):
     if ancestral_allele is None:
          # store tskit.MISSING_DATA internally
     # do the stuff.
```

Reply:
Having ``ancestral_allele=-1`` as missing data seems most logical now that we're making it an integer? It's good to reserve ``None`` as "I didn't specify anything". Otherwise, we won't know if the user didn't specify anything for ancestral_allele or whether they specifically said "it must be 0".

Reply:
The tskit.MISSING_DATA isn't particularly helpful here I think - sgkit also uses -1 as missing data, and that's a general rule that a user is going to need to understand.

Reply:
As long as we are happy with `-1` not being used by the user to mean the last allele in the list. I'm not sure how we would catch that.

Reply:
That's just a general gotcha, can't be helped.